File: software/examples/c/example_modbus_master.c
#include <stdio.h>

#include "ip_connection.h"
#include "bricklet_rs485.h"

#define HOST "localhost"
#define PORT 4223
#define UID "XYZ" // Change XYZ to the UID of your RS232 Bricklet

RS485 rs485;
uint8_t expected_request_id = 0;

// Callback function for write single register response callback
void cb_modbus_master_write_single_register_response(uint8_t request_id,
                                                     int8_t exception_code,
                                                     void *user_data) {
    if(request_id != expected_request_id) {
      printf("Unexpected request ID\n");

      return;
    }

    printf("Request ID = %d\n", request_id);
    printf("Exception Code = %d\n", exception_code);
}

int main(void) {
    // Create IP connection
    IPConnection ipcon;
    ipcon_create(&ipcon);

    // Create device object
    rs485_create(&rs485, UID, &ipcon);

    // Connect to brickd
    if(ipcon_connect(&ipcon, HOST, PORT) < 0) {
        fprintf(stderr, "Could not connect\n");

        return 1;
    }
    // Don't use device before ipcon is connected

    // Set operating mode of the Bricklet
    rs485_set_mode(&rs485, RS485_MODE_MODBUS_MASTER_RTU);

    /*
     * Modbus specific configuration
     *
     * Slave mode address = 1 (Unused in master mode)
     * Master mode request timeout = 1000ms (Unused in slave mode)
     */
    rs485_set_modbus_configuration(&rs485, 1, 1000);

    // Register write single register response callback
    rs485_register_callback(&rs485,
                            RS485_CALLBACK_MODBUS_MASTER_WRITE_SINGLE_REGISTER_RESPONSE,
                            (void *)cb_modbus_master_write_single_register_response,
                            NULL);

    // Request single register write
    rs485_modbus_master_write_single_register(&rs485, 1, 42, 65535, &expected_request_id);

    fprintf(stdout, "Press key to exit\n");
    getchar();

    rs485_destroy(&rs485);
    ipcon_destroy(&ipcon); // Calls ipcon_disconnect internally

    return 0;
}

File: software/examples/go/example_modbus_master.go
package main

import (
	"fmt"
	"github.com/Tinkerforge/go-api-bindings/ipconnection"
    "github.com/Tinkerforge/go-api-bindings/rs485_bricklet"
)

const ADDR string = "localhost:4223"
const UID string = "XYZ" // Change XYZ to the UID of your RS485 Bricklet.

func main() {
	ipcon := ipconnection.New()
    defer ipcon.Close()
	rs485, _ := rs485_bricklet.New(UID, &ipcon) // Create device object.

	ipcon.Connect(ADDR) // Connect to brickd.
    defer ipcon.Disconnect()
	// Don't use device before ipcon is connected.

	// Set operating mode to Modbus RTU master
	rs485.SetMode(rs485_bricklet.ModeModbusMasterRTU)

	// Modbus specific configuration:
	// - slave address = 1 (unused in master mode)
	// - master request timeout = 1000ms
	rs485.SetModbusConfiguration(1, 1000)

    expectedRequestID := uint8(0)
    
	rs485.RegisterModbusMasterWriteSingleRegisterResponseCallback(func(requestID uint8, exceptionCode rs485_bricklet.ExceptionCode) {
		fmt.Printf("Request ID: %d", requestID);
        if exceptionCode == rs485_bricklet.ExceptionCodeTimeout {
			fmt.Println("Exception Code: Timeout")
		} else if exceptionCode == rs485_bricklet.ExceptionCodeSuccess {
			fmt.Println("Exception Code: Success")
		} else if exceptionCode == rs485_bricklet.ExceptionCodeIllegalFunction {
			fmt.Println("Exception Code: Illegal Function")
		} else if exceptionCode == rs485_bricklet.ExceptionCodeIllegalDataAddress {
			fmt.Println("Exception Code: Illegal Data Address")
		} else if exceptionCode == rs485_bricklet.ExceptionCodeIllegalDataValue {
			fmt.Println("Exception Code: Illegal Data Value")
		} else if exceptionCode == rs485_bricklet.ExceptionCodeSlaveDeviceFailure {
			fmt.Println("Exception Code: Slave Device Failure")
		} else if exceptionCode == rs485_bricklet.ExceptionCodeAcknowledge {
			fmt.Println("Exception Code: Acknowledge")
		} else if exceptionCode == rs485_bricklet.ExceptionCodeSlaveDeviceBusy {
			fmt.Println("Exception Code: Slave Device Busy")
		} else if exceptionCode == rs485_bricklet.ExceptionCodeMemoryParityError {
			fmt.Println("Exception Code: Memory Parity Error")
		} else if exceptionCode == rs485_bricklet.ExceptionCodeGatewayPathUnavailable {
			fmt.Println("Exception Code: Gateway Path Unavailable")
		} else if exceptionCode == rs485_bricklet.ExceptionCodeGatewayTargetDeviceFailedToRespond {
			fmt.Println("Exception Code: Gateway Target Device Failed To Respond")
		}
        if requestID != expectedRequestID {
            fmt.Println("Error: Unexpected request ID");
        }
	})

	// Write 65535 to register 42 of slave 17
	expectedRequestID, _ = rs485.ModbusMasterWriteSingleRegister(17, 42, 65535)

	fmt.Print("Press enter to exit.")
	fmt.Scanln()

	ipcon.Disconnect()
}
